pkg/sync/kubernetes/kubernetes_sync.go

```diff
@@ -72,36 +88,121 @@ func (k *Sync) Notify(ctx context.Context, c chan<- sync.INotify) {
 
-	go featureflagconfiguration.WatchResources(ctx, *k.Logger.WithFields(log.Fields{
-		"sync":      "kubernetes",
-		"component": "watchresources",
-	}), k.client, resyncPeriod, controllerClient.ObjectKey{
+	objectKey := client.ObjectKey{
 		Name:      k.ProviderArgs[featureFlagConfigurationName],
 		Namespace: k.ProviderArgs[featureFlagNamespaceName],
-	}, c)
+	}
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if err := createFuncHandler(obj, objectKey, c); err != nil {
+				k.Logger.Warn(err.Error())
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			if err := updateFuncHandler(oldObj, newObj, objectKey, c); err != nil {
+				k.Logger.Warn(err.Error())
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			if err := deleteFuncHandler(obj, objectKey, c); err != nil {
+				k.Logger.Warn(err.Error())
+			}
+		},
+	})
+
+	informer.Run(ctx.Done())
+}
+
+func createFuncHandler(obj interface{}, object client.ObjectKey, c chan<- sync.INotify) error {
+	var ffObj v1alpha1.FeatureFlagConfiguration
+	u := obj.(*unstructured.Unstructured)
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &ffObj)
+	if err != nil {
+		return err
+	}
+	if ffObj.APIVersion != v1alpha1.GroupVersion.Version {
+		return errors.New("invalid api version")
+	}
+	if ffObj.Name == object.Name {
+		c <- &sync.Notifier{
+			Event: sync.Event[sync.DefaultEventType]{
+				EventType: sync.DefaultEventTypeCreate,
+			},
+		}
+	}
+	return nil
+}
+
+func updateFuncHandler(oldObj interface{}, newObj interface{}, object client.ObjectKey, c chan<- sync.INotify) error {
+	var ffOldObj v1alpha1.FeatureFlagConfiguration
+	u := oldObj.(*unstructured.Unstructured)
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &ffOldObj)
+	if err != nil {
+		return err
+	}
+	if ffOldObj.APIVersion != v1alpha1.GroupVersion.Version {
+		return errors.New("invalid api version")
+	}
+	var ffNewObj v1alpha1.FeatureFlagConfiguration
+	u = newObj.(*unstructured.Unstructured)
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &ffNewObj)
+	if err != nil {
+		return err
+	}
+	if ffNewObj.APIVersion != v1alpha1.GroupVersion.Version {
+		return errors.New("invalid api version")
+	}
+	if object.Name == ffNewObj.Name && ffOldObj.ResourceVersion != ffNewObj.ResourceVersion {
+		// Only update if there is an actual featureFlagSpec change
+		c <- &sync.Notifier{
+			Event: sync.Event[sync.DefaultEventType]{
+				EventType: sync.DefaultEventTypeModify,
+			},
+		}
+	}
+	return nil
+}
+
+func deleteFuncHandler(obj interface{}, object client.ObjectKey, c chan<- sync.INotify) error {
+	var ffObj v1alpha1.FeatureFlagConfiguration
+	u := obj.(*unstructured.Unstructured)
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &ffObj)
+	if err != nil {
+		return err
+	}
+	if reflect.TypeOf(ffObj) != reflect.TypeOf(v1alpha1.FeatureFlagConfiguration{}) {
```

Comment:
This `if` statement will never fail, it should be updated to match the pattern in `updateFuncHandler` and `createFuncHandler`